### ISSUE TITLE
withArgs matches incorrectly with multiple react contexts

### ISSUE THREAD

**Issue by J-France** [sinonjs/sinon#2286]
**Describe the bug**
When using `withArgs` with react contexts, the return appears to always match the final context added.

**To Reproduce**
```js
import React from 'react';
import sinon from 'sinon';

const { createContext } = React;

const context1 = createContext();
const context2 = createContext();

const stub = sinon.stub()
stub.withArgs(context1).returns(1);
stub.withArgs(context2).returns(2);

console.log(stub(context1));
console.log(stub(context2));
console.log(Object.is(context1, context2));
```

**Expected behavior**
The first console.log should be a 1. However both logs are producing a 2.

**Context (please complete the following information):**

* Library version:
sinon: 9.0.3
* Environment:
node: v14.8.0
* Other libraries you are using:
react: 16.13.1

**Additional context**

**Comment by mantoni** [comment]
`stub.withArgs` uses the `deepEqual` function from [@sinonjs/samsam](https://github.com/sinonjs/samsam) to find matching behaviors. I suppose two different React context instances look the same to this algorithm.

To make this work, we could try to find matching behaviors in two phases:

1. Use a strict `===` check on all arguments to find matching fakes.
2. Fall back to the current implementation if (1) doesn't match anything.

The matching is implemented here:

https://github.com/sinonjs/sinon/blob/34d921042ea8c58c66d7e16187436f07e59bb898/lib/sinon/spy.js#L24-L30

Would you be interested in contributing a fix?

**Comment by J-France** [comment]
I took a quick look and I just want to be sure I understand the change you are proposing before going any further.

The `withArgs` call would need to be altered to only use reference equality (so that it doesn't return the first fake right away when doing the second `withArgs` call) and then use a combination of reference (checked first) and value equality (if not reference equal) when doing the actual stub call?

**Comment by mantoni** [comment]
We need to retain the existing comparison logic and cannot switch to strict equality. We could find matching behaviors in two passes, first by strict equality and if that does not return any matches, use the existing implementation. This would fix your use case.

**Comment by J-France** [comment]
If you retain deep equality on the withArgs call it wouldn't work because you would not be able to assign different return values
```js
const obj1 = { A: 1 }
const obj2 = { A:1 }
const stub = sinon.stub()
const fake1 = stub.withArgs(obj1) // creates a new fake
const fake2 = stub.withArgs(obj2) // deep equals obj1 so returns the same fake
Object.is(fake1, fake2) // true
```
if you switch back to `===` equality for it, you would end up with two different fakes and it would test them both when you called the stub. I believe this should maintain existing behavior and provide a 'priority' match for reference equality but I do not know enough to say what potential side effects this could lead to or if there are parts of the code which would not support it.

If this is not a reasonable change to make, I would consider adding a note to the `withArgs` doc similar to the one for `returned` sufficient as it was easy enough for me to work around it once I understood what was happening.
```
Uses deep comparison for objects and arrays. Use spy.withArgs(sinon.match.same(obj)) for strict comparison (see matchers).
```

**Comment by mantoni** [comment]
You suggestion to update the documentation is probably the better option. Using `sinon.match.same(obj)` makes clear what is going on in the test case as well, rather than having some magic to find the correct behavior.

Would you care to prepare a pull request with the documentation change?

**Comment by mantoni** [comment]
Solved with a documentation improvement in #2290.
